Fix the memory leak described in Issue #33178

Tetragramm opened the pull request:
Documented in [THIS](https://github.com/tensorflow/tensorflow/issues/33178) thread.
Based on the documentation of get() and generic_utils.object_list_uid, this has no functional effect, except to remove an unnecessary map that was growing with every input.

Tested using the example program in [THIS](https://github.com/tensorflow/tensorflow/issues/33178#issuecomment-541971922) post.  That could be turned into a unit-test, but the actual amount of memory before it fails is dependent on GPU, and I'm not sure what guidelines or methods there are to mitigate that.

Tetragramm committed:
Fix the memory leak described in Issue #33178

gbaned commented:
@Tetragramm Can you please resolve conflicts? Thanks!

Tetragramm committed:
Merge remote-tracking branch 'remotes/upstream/master' into Issue_#33178

# Conflicts:
#	tensorflow/python/keras/layers/wrappers.py

Tetragramm commented:
Ok, that should do it.

qlzh727 commented on tensorflow/python/keras/layers/wrappers.py:
```diff
@@ -318,9 +314,7 @@ def compute_mask(self, inputs, mask=None):
-    input_uid = generic_utils.object_list_uid(inputs)
-    inner_inputs = self._input_map.get(input_uid, inputs)
-    output_mask = self.layer.compute_mask(inner_inputs, inner_mask)
+    output_mask = self.layer.compute_mask(inputs, inner_mask)
```
The input might have a different shape since it was reshaped in the call().

```
inner_input_shape = self._get_shape_tuple((-1,), inputs, 2)
# Shape: (num_samples * timesteps, ...). And track the	        
inputs = array_ops.reshape(inputs, inner_input_shape)
```

Tetragramm replied:
I see the problem there.
Memory usage would be greatly reduced by storing shape instead of the actual inputs and just reshaping again.  That doesn't actually fix the problem though.  

Does compute_mask get called always once, and only once?  It could just remove it from the dict if so, but I can't find anything specific on that.

qlzh727 replied:
compute_mask is a public API, so it can be called in all different situation, eg two masks can be created if the same layer is called with two different inputs.

Tetragramm replied:
All the reshape appears to be the same every time, and I don't see any dependencies on anything but the variable "input".

Which means we could perform the same reshape in compute mask. Did I miss anything?

qlzh727 replied:
The existing code was trying to get the inner_inputs if it is there, which might be reshaped compare to original inputs.

If we want to want cache the inputs, we somehow need to reshape the inputs again, and make sure the layer.compute_mask() is called with same input tensor when layer.call() is invoked.

Tetragramm replied:
Hmm. I think compute_mask() was not working quite right already.  If you'd already called call(), it would have a cached value and use the reshaped input.  If you hadn't, it would just use the original input.

I'm not sure which is the correct behavior.  I'm would think the reshaped input is correct.

qlzh727 replied:
Sorry for the long wait. Thinking about this more, I think its actually better to remove the unbounded caching here with a small dip for the performance.

qlzh727 commented:
Also adding Francois, since this caching field was added a long times ago.

Tetragramm committed:
Reshape input in compute_mask(), the same way it is done in call().
This should suffice in place of saving the shape in the dict that was causing the leak.

gbaned commented:
@qlzh727 Any update on this PR, please. Thanks!

qlzh727 reviewed (approved):
Please sync/rebase you change, and I will kick off the test again. Sorry for the long wait.

Tetragramm committed:
Merge branch 'master' of https://github.com/tensorflow/tensorflow into Issue_#33178

Tetragramm commented:
It should be ready to be tested.

gbaned commented:
@Tetragramm Can you please address Ubuntu Sanity errors? Thanks!

Tetragramm committed:
FAIL: Found 1 non-whitelisted pylint errors:
tensorflow/python/keras/layers/wrappers.py:267: [C0330(bad-continuation), ] Wrong continued indentation (remove 1 space).

Removed 1 space.

Tetragramm commented:
They should be fixed, but I guess you have to kick off the checks?  They didn't run, anyway.

Tetragramm commented:
The status says an error happened.  I don't see any details.  Is there anything I need to do?  Rebase the change again?

gbaned commented:
@Tetragramm It is waiting for a import internally. We will let you know if anything is required from you. Thank you.
